File: src/update.ts
/** Shared GitHub latest-release fetch for `hwf update` and the picker indicator. */

import { chdir } from "node:process";
import { homedir, tmpdir } from "node:os";
import { join } from "node:path";
import { die } from "./herdr";
import { resolvePluginRoot } from "./setup";
import { PRODUCT_VERSION } from "./version";

const RELEASE_REPO = "aorumbayev/herdr-workflows";
const LATEST_RELEASE_URL = `https://api.github.com/repos/${RELEASE_REPO}/releases/latest`;
const DEFAULT_RELEASE_CHECK_TIMEOUT_MS = 8_000;

export type LatestRelease = {
  tag: string;
  version: string;
};

export class ReleaseCheckError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ReleaseCheckError";
  }
}

/** Strict `v0.x.y` tag → bare `0.x.y` version. */
export function parseReleaseTag(tag: string): LatestRelease {
  const m = /^v(0\.\d+\.\d+)$/.exec(tag.trim());
  if (!m) {
    throw new ReleaseCheckError(
      `latest release tag is not a strict v0.x.y semver: ${JSON.stringify(tag)}`,
    );
  }
  return { tag: `v${m[1]}`, version: m[1]! };
}

/** Compare `0.x.y` versions. Negative when a < b. */
export function compareSemver(a: string, b: string): number {
  const pa = parseParts(a);
  const pb = parseParts(b);
  for (let i = 0; i < 3; i++) {
    if (pa[i]! !== pb[i]!) return pa[i]! < pb[i]! ? -1 : 1;
  }
  return 0;
}

function parseParts(version: string): [number, number, number] {
  const m = /^(0)\.(\d+)\.(\d+)$/.exec(version.trim());
  if (!m) throw new ReleaseCheckError(`expected 0.x.y version, got ${JSON.stringify(version)}`);
  return [Number(m[1]), Number(m[2]), Number(m[3])];
}

export type FetchLatestOptions = {
  timeoutMs?: number;
  url?: string;
  fetchImpl?: (url: string, init?: RequestInit) => Promise<Response>;
};

/**
 * Fetch only the latest *published* GitHub Release (drafts are not `/releases/latest`).
 * Network/parse failures throw ReleaseCheckError.
 */
export async function fetchLatestPublishedRelease(
  opts: FetchLatestOptions = {},
): Promise<LatestRelease> {
  const timeoutMs = opts.timeoutMs ?? DEFAULT_RELEASE_CHECK_TIMEOUT_MS;
  const url = opts.url ?? LATEST_RELEASE_URL;
  const fetchImpl = opts.fetchImpl ?? fetch;
  const ac = new AbortController();
  const timer = setTimeout(() => ac.abort(), timeoutMs);
  try {
    const res = await fetchImpl(url, {
      signal: ac.signal,
      headers: {
        Accept: "application/vnd.github+json",
        "User-Agent": "herdr-workflows",
        "X-GitHub-Api-Version": "2022-11-28",
      },
    });
    if (!res.ok) {
      throw new ReleaseCheckError(`latest release request failed: HTTP ${res.status}`);
    }
    const body = (await res.json()) as { tag_name?: unknown; draft?: unknown };
    if (body.draft === true) {
      throw new ReleaseCheckError("latest release endpoint returned a draft");
    }
    if (typeof body.tag_name !== "string") {
      throw new ReleaseCheckError("latest release response missing tag_name");
    }
    return parseReleaseTag(body.tag_name);
  } catch (error) {
    if (error instanceof ReleaseCheckError) throw error;
    if (error instanceof Error && error.name === "AbortError") {
      throw new ReleaseCheckError(`latest release request timed out after ${timeoutMs}ms`);
    }
    throw new ReleaseCheckError(
      `latest release request failed: ${error instanceof Error ? error.message : String(error)}`,
    );
  } finally {
    clearTimeout(timer);
  }
}

export type PluginSourceInfo = {
  kind: "github" | "local" | "unregistered";
  owner?: string;
  repo?: string;
};

export type UpdateDeps = {
  fetchLatest?: (opts?: FetchLatestOptions) => Promise<{ version: string; tag: string }>;
  runInstall?: (args: string[], cwd: string) => Promise<number>;
};

/**
 * `hwf update` — check latest published release and delegate replacement to Herdr.
 * Must not import the picker module.
 */
export async function runUpdate(deps: UpdateDeps = {}): Promise<void> {
  const fetchLatest = deps.fetchLatest ?? fetchLatestPublishedRelease;
  const runInstall = deps.runInstall ?? runHerdrInstall;

  let latest: { version: string; tag: string };
  try {
    latest = await fetchLatest();
  } catch (error) {
    const msg = error instanceof ReleaseCheckError ? error.message : String(error);
    die(`update check failed: ${msg}`);
  }

  if (compareSemver(PRODUCT_VERSION, latest.version) >= 0) {
    process.stdout.write(`already up to date (${PRODUCT_VERSION})\n`);
    return;
  }

  const source = await resolvePluginSource();
  if (source.kind === "local") {
    die(
      `refusing to update a linked development checkout — run bun run install:dev from the working tree instead`,
    );
  }
  if (source.kind === "unregistered") {
    die(
      `this binary is not a Herdr-managed herdr-workflows install — run: herdr plugin install ${RELEASE_REPO}`,
    );
  }

  process.stdout.write(
    `updating ${PRODUCT_VERSION} → ${latest.version} via herdr plugin install ${RELEASE_REPO}\n`,
  );
  const root = resolvePluginRoot();
  const cwd = leavePluginRoot(root);
  const code = await runInstall(
    ["plugin", "install", RELEASE_REPO, "--ref", latest.tag, "--yes"],
    cwd,
  );
  if (code !== 0) {
    process.stderr.write(`herdr plugin install failed with exit ${code}\n`);
    process.exit(code);
  }
  process.stdout.write(`updated to ${latest.version}\n`);
}

export function leavePluginRoot(
  pluginRootPath: string,
  env: NodeJS.ProcessEnv = process.env,
): string {
  const candidates = [homedir(), tmpdir(), env.HOME].filter((p): p is string =>
    Boolean(p && p.length > 0),
  );
  const normalizedRoot = join(pluginRootPath);
  for (const candidate of candidates) {
    const abs = join(candidate);
    if (abs !== normalizedRoot && !abs.startsWith(normalizedRoot + "/")) {
      try {
        chdir(abs);
        return abs;
      } catch {
        // try next
      }
    }
  }
  // Last resort: parent of plugin root when it is not the filesystem root.
  const parent = join(pluginRootPath, "..");
  if (parent !== normalizedRoot) {
    chdir(parent);
    return parent;
  }
  throw new Error(`cannot leave HERDR_PLUGIN_ROOT ${pluginRootPath}`);
}

async function resolvePluginSource(
  env: NodeJS.ProcessEnv = process.env,
): Promise<PluginSourceInfo> {
  const herdr = env.HERDR_BIN_PATH?.trim() || "herdr";
  const proc = Bun.spawn([herdr, "plugin", "list", "--json", "--plugin", "herdr-workflows"], {
    stdout: "pipe",
    stderr: "pipe",
    env,
  });
  const [stdout, stderr, code] = await Promise.all([
    new Response(proc.stdout).text(),
    new Response(proc.stderr).text(),
    proc.exited,
  ]);
  if (code !== 0) {
    // Missing plugin / herdr unavailable → treat as unregistered for clear guidance.
    if (/not found|no such|unknown plugin/i.test(stdout + stderr)) {
      return { kind: "unregistered" };
    }
    throw new Error(`herdr plugin list failed: ${(stderr || stdout).trim() || `exit ${code}`}`);
  }
  return parsePluginListSource(stdout);
}

export function parsePluginListSource(jsonText: string): PluginSourceInfo {
  let parsed: unknown;
  try {
    parsed = JSON.parse(jsonText);
  } catch {
    throw new Error("herdr plugin list returned invalid JSON");
  }
  const plugin = findHerdrWorkflowsPlugin(parsed);
  if (!plugin) return { kind: "unregistered" };
  const source = plugin.source;
  if (!source || typeof source !== "object") return { kind: "local" };
  const kind = (source as { kind?: unknown }).kind;
  if (kind === "github") {
    const owner = str((source as { owner?: unknown }).owner);
    const repo = str((source as { repo?: unknown }).repo);
    return { kind: "github", owner, repo };
  }
  return { kind: "local" };
}

/**
 * Herdr CLI `--json` prints `{ id, result: { type: "plugin_list", plugins: [...] } }`.
 * Offline/error envelopes and a bare `result` object are accepted for the same shape.
 */
function findHerdrWorkflowsPlugin(parsed: unknown): { source?: unknown } | undefined {
  if (!parsed || typeof parsed !== "object") return undefined;
  const obj = parsed as Record<string, unknown>;
  const result =
    obj.result && typeof obj.result === "object"
      ? (obj.result as Record<string, unknown>)
      : obj.type === "plugin_list"
        ? obj
        : undefined;
  if (!result || result.type !== "plugin_list" || !Array.isArray(result.plugins)) {
    return undefined;
  }
  for (const entry of result.plugins) {
    if (!entry || typeof entry !== "object") continue;
    const id = (entry as { plugin_id?: unknown }).plugin_id;
    if (id === "herdr-workflows") return entry as { source?: unknown };
  }
  return undefined;
}

function str(value: unknown): string | undefined {
  return typeof value === "string" && value.length > 0 ? value : undefined;
}

async function runHerdrInstall(args: string[], cwd: string): Promise<number> {
  const herdr = process.env.HERDR_BIN_PATH?.trim() || "herdr";
  const proc = Bun.spawn([herdr, ...args], {
    cwd,
    stdout: "inherit",
    stderr: "inherit",
    stdin: "inherit",
    env: process.env,
  });
  return proc.exited;
}

File: internal/cli/update.go
package cli

import (
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"regexp"
	"strings"

	assets "github.com/aorumbayev/herdr-workflows/embed"
	"github.com/aorumbayev/herdr-workflows/internal/host"
	"github.com/aorumbayev/herdr-workflows/internal/update"
)

type exitCodeError struct {
	code int
	msg  string
}

func (e *exitCodeError) Error() string { return e.msg }
func (e *exitCodeError) ExitCode() int { return e.code }

type updateDeps struct {
	FetchLatest    func() (update.LatestRelease, error)
	RunInstall     func(args []string, cwd string) (int, error)
	ListSource     func() (update.PluginSourceInfo, error)
	PluginRoot     string
	Version        string
	Getenv         func(string) string
	Executable     func() (string, error)
	InstallRelease func(update.InstallOpts) error
}

var pluginListMissingRE = regexp.MustCompile(`(?i)not found|no such|unknown plugin`)

func resolvePluginSource(getenv func(string) string) (update.PluginSourceInfo, error) {
	if getenv == nil {
		getenv = os.Getenv
	}
	herdr := host.BinPath(getenv)
	cmd := exec.Command(herdr, "plugin", "list", "--json", "--plugin", "herdr-workflows")
	cmd.Env = os.Environ()
	var stderrBuf strings.Builder
	cmd.Stderr = &stderrBuf
	out, err := cmd.Output()
	if err == nil {
		src, parseErr := update.ParsePluginListSource(string(out))
		if parseErr != nil {
			return update.PluginSourceInfo{}, parseErr
		}
		return src, nil
	}
	var exitErr *exec.ExitError
	if !errors.As(err, &exitErr) {
		return update.PluginSourceInfo{}, fmt.Errorf("herdr plugin list failed: %s", err.Error())
	}
	combined := stderrBuf.String() + string(out)
	if pluginListMissingRE.MatchString(combined) {
		return update.PluginSourceInfo{Kind: "unregistered"}, nil
	}
	msg := strings.TrimSpace(combined)
	if msg == "" {
		msg = fmt.Sprintf("exit %d", exitErr.ExitCode())
	}
	return update.PluginSourceInfo{}, fmt.Errorf("herdr plugin list failed: %s", msg)
}

func defaultHerdrInstall(getenv func(string) string) func([]string, string) (int, error) {
	return func(args []string, cwd string) (int, error) {
		herdr := host.BinPath(getenv)
		cmd := exec.Command(herdr, args...)
		cmd.Dir = cwd
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		cmd.Stdin = os.Stdin
		cmd.Env = os.Environ()
		err := cmd.Run()
		if err == nil {
			return 0, nil
		}
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			return exitErr.ExitCode(), nil
		}
		return 0, err
	}
}

func pluginRootFor(getenv func(string) string) string {
	execPath, _ := os.Executable()
	cwd, _ := os.Getwd()
	return resolvePluginRoot(getenv, execPath, cwd)
}

func defaultUpdateDeps() updateDeps {
	getenv := os.Getenv
	return updateDeps{
		FetchLatest:    func() (update.LatestRelease, error) { return update.CheckForUpdate(update.CheckOpts{}) },
		RunInstall:     defaultHerdrInstall(getenv),
		ListSource:     func() (update.PluginSourceInfo, error) { return resolvePluginSource(getenv) },
		PluginRoot:     pluginRootFor(getenv),
		Version:        assets.ManifestVersion(),
		Getenv:         getenv,
		Executable:     os.Executable,
		InstallRelease: update.InstallRelease,
	}
}

func refVersionFromInstallArgs(args []string) string {
	for i := 0; i < len(args)-1; i++ {
		if args[i] == "--ref" {
			tag := strings.TrimPrefix(args[i+1], "v")
			return tag
		}
	}
	return ""
}

func executeUpdate(deps updateDeps, stdout, stderr io.Writer) error {
	if deps.Getenv == nil {
		deps.Getenv = os.Getenv
	}
	if deps.FetchLatest == nil {
		deps.FetchLatest = func() (update.LatestRelease, error) { return update.CheckForUpdate(update.CheckOpts{}) }
	}
	if deps.ListSource == nil {
		deps.ListSource = func() (update.PluginSourceInfo, error) { return resolvePluginSource(deps.Getenv) }
	}
	if deps.RunInstall == nil {
		deps.RunInstall = defaultHerdrInstall(deps.Getenv)
	}
	if deps.Executable == nil {
		deps.Executable = os.Executable
	}
	if deps.InstallRelease == nil {
		deps.InstallRelease = update.InstallRelease
	}
	if deps.PluginRoot == "" {
		deps.PluginRoot = pluginRootFor(deps.Getenv)
	}
	current := deps.Version
	if current == "" {
		current = assets.ManifestVersion()
	}
	origInstall := deps.RunInstall
	deps.RunInstall = func(args []string, cwd string) (int, error) {
		to := refVersionFromInstallArgs(args)
		if to == "" {
			to = "?"
		}
		if _, err := fmt.Fprintf(stdout, "updating %s → %s via herdr plugin install %s\n", current, to, update.ReleaseRepo); err != nil {
			return 0, err
		}
		return origInstall(args, cwd)
	}
	origStandalone := deps.InstallRelease
	deps.InstallRelease = func(opts update.InstallOpts) error {
		if _, err := fmt.Fprintf(stdout, "updating %s → %s via standalone binary replace\n", current, opts.Version); err != nil {
			return err
		}
		return origStandalone(opts)
	}
	result, err := update.UpdatePlugin(update.Deps{
		FetchLatest:    deps.FetchLatest,
		RunInstall:     deps.RunInstall,
		ListSource:     deps.ListSource,
		PluginRoot:     deps.PluginRoot,
		Version:        current,
		Getenv:         deps.Getenv,
		Executable:     deps.Executable,
		InstallRelease: deps.InstallRelease,
	})
	if err != nil {
		msg := err.Error()
		var checkErr *update.ReleaseCheckError
		if errors.As(err, &checkErr) {
			return fmt.Errorf("update check failed: %s", msg)
		}
		return fmt.Errorf("update failed: %s", msg)
	}
	return presentUpdateResult(result, stdout, stderr)
}

func presentUpdateResult(result update.Result, stdout, stderr io.Writer) error {
	switch result.Kind {
	case "up_to_date":
		if _, err := fmt.Fprintf(stdout, "already up to date (%s)\n", result.Current); err != nil {
			return err
		}
		return nil
	case "refused_local":
		return fmt.Errorf("refusing to update a linked development checkout — run go run ./scripts/install-dev from the working tree instead")
	case "refused_unregistered":
		return fmt.Errorf("standalone update refused")
	case "updated":
		if _, err := fmt.Fprintf(stdout, "updated to %s\n", result.To); err != nil {
			return err
		}
		return nil
	case "install_failed":
		msg := fmt.Sprintf("herdr plugin install failed with exit %d", result.Code)
		if _, err := fmt.Fprintln(stderr, msg); err != nil {
			return err
		}
		return &exitCodeError{code: result.Code, msg: msg}
	default:
		return fmt.Errorf("update failed: unknown result kind %q", result.Kind)
	}
}
